### ISSUE TITLE
LLDB build problems on s390x with recent glibc

### ISSUE THREAD

**Issue by tbaederr** [llvm/llvm-project#56280]
lldb fails to build on s390x with a recent glibc:

```
In file included from /builddir/build/BUILD/lldb-14.0.5.src/source/Plugins/Process/Linux/NativeRegisterContextLinux_s390x.cpp:21:
/usr/include/linux/uio.h:17:8: error: redefinition of 'struct iovec'
   17 | struct iovec
      |        ^~~~~
In file included from /usr/include/bits/mman_ext.h:24,
                 from /usr/include/sys/mman.h:150,
                 from /builddir/build/BUILD/lldb-14.0.5.src/source/Plugins/Process/Linux/IntelPTManager.h:19,
                 from /builddir/build/BUILD/lldb-14.0.5.src/source/Plugins/Process/Linux/NativeProcessLinux.h:23,
                 from /builddir/build/BUILD/lldb-14.0.5.src/source/Plugins/Process/Linux/NativeRegisterContextLinux_s390x.cpp:12:
/usr/include/bits/types/struct_iovec.h:26:8: note: previous definition of 'struct iovec'
   26 | struct iovec
      |        ^~~~~
```

`struct iovec` is defined both in `/usr/include/bits/types/struct_iovec.h` (from glibc) and `/usr/include/linux/uio.h` (from the kernel).

The problem only exists in s390x. FWIW I've found that just removing the `<linux/uio.h>` include from `NativeRegisterContextLinux_s390x.cpp` seems to fix the build.

**Comment by fweimer-rh** [comment]
The problem appears to be s390x-specific because the source file is.

The source of the incompatibility is the addition of `process_madvise` to `<sys/mman.h>`, which also brought in a definition of `struct iovec`. We should probably merely use a forward declaration in this context.

**Comment by labath** [comment]
> FWIW I've found that just removing the <linux/uio.h> include from NativeRegisterContextLinux_s390x.cpp seems to fix the build.

That would make the build fail (due to undefined struct iovec) on older libcs right?

Mixing libc and linux headers is known to cause issues, and I don't think there's a good reason why this code needs to include the latter would replacing linux/uio.h with sys/uio.h (which is a libc header I believe) solve the issue?

**Comment by fweimer-rh** [comment]
glibc patch posted: [[PATCH] Linux: Forward declaration of struct iovec for process_madvise](https://sourceware.org/pipermail/libc-alpha/2022-June/140172.html)

**Comment by fweimer-rh** [comment]
> Mixing libc and linux headers is known to cause issues, and I don't think there's a good reason why this code needs to include the latter. Would replacing linux/uio.h with sys/uio.h (which is a libc header I believe) solve the issue?

@labath Yes, that is a good idea in general. We will likely fix this on the glibc side, too, but if you can avoid mixing Linux and glibc headers easily, it's always a good thing to do that.
